src/main/java/com/wanderersoftherift/wotr/world/level/levelgen/processor/util/ProcessorUtil.java

```diff
@@ -0,0 +1,209 @@
+            return block == null || block.defaultBlockState().isAir() ;
+        }else {
+            return blockinfo == null || blockinfo.state.is(AIR) || blockinfo.state.is(CAVE_AIR);
+        }
+    }*/
+
+    public static boolean isSolid(StructureTemplate.StructureBlockInfo blockinfo){
+        if(blockinfo != null && blockinfo.state().is(JIGSAW)){
+            Block block = BuiltInRegistries.BLOCK.getValue(ResourceLocation.parse(blockinfo.nbt().getString(NBT_FINAL_STATE)));
+            return block != null && !block.defaultBlockState().is(AIR) && !(block instanceof LiquidBlock);
+        }else {
+            return blockinfo != null && !blockinfo.state().is(AIR) && !(blockinfo.state().getBlock() instanceof LiquidBlock);
+        }
+    }
+
+    public static boolean isFaceFull(StructureTemplate.StructureBlockInfo blockinfo, Direction direction) {
+        if(blockinfo == null) return false;
+        if (blockinfo.state().is(JIGSAW)) {
+            Block block = BuiltInRegistries.BLOCK.getValue(ResourceLocation.parse(blockinfo.nbt().getString(NBT_FINAL_STATE)));
+            return block != null && !blockinfo.state().is(AIR) && !(blockinfo.state().getBlock() instanceof LiquidBlock) &&
+                    Block.isFaceFull(((InvokerBlockBehaviour) block).invokeGetShape(block.defaultBlockState(), null, blockinfo.pos(), CollisionContext.empty()), direction);
+        } else {
+            return !blockinfo.state().is(AIR) && !(blockinfo.state().getBlock() instanceof LiquidBlock) && Block.isFaceFull(((InvokerBlockBehaviour) blockinfo.state().getBlock()).invokeGetCollisionShape(blockinfo.state(), null, blockinfo.pos(), CollisionContext.empty()), direction);
+        }
+    }
+
+    public static StructureTemplate.Palette getCurrentPalette(List<StructureTemplate.Palette> palettes) {
+        int i = palettes.size();
+        if (i == 0) {
+            throw new IllegalStateException("No palettes");
+        } else {
+            return palettes.get(i - 1);
+        }
+    }
+
+    public static StructureTemplate.StructureBlockInfo getBlockInfo(List<StructureTemplate.StructureBlockInfo> mapByPos, BlockPos pos) {
+        BlockPos firstPos = mapByPos.getFirst().pos();
+        BlockPos lastPos = mapByPos.getLast().pos();
+        if(isPosBetween(pos, firstPos, lastPos)){
+            return null;
+        }
+        int width = lastPos.getX() + 1 - firstPos.getX();
+        int height = lastPos.getY() + 1 - firstPos.getY();
+        int index = (pos.getX() - firstPos.getX()) + (pos.getY()- firstPos.getY()) * width + (pos.getZ()-firstPos.getZ()) * width * height;
+        if(index < 0 || index >= mapByPos.size()){
+            return null;
+        }
+        return mapByPos.get(index);
+    }
+
+    private static boolean isPosBetween(BlockPos pos, BlockPos firstPos, BlockPos lastPos) {
+        return pos.getX() < firstPos.getX() || pos.getX() > lastPos.getX() ||
```

Comment:
I feel the logic here is inverted, should wrap the whole result with `!` (e.g. if pos.getX() is less than firstPos.getX() then the method returns `true` currently, but should be `false`).

Reply:
Or otherwise the method name is misleading.

Reply:
Flipped it. Now checks ands